File: BST/lua/node.lua

Node = {}

function Node.new (value)
    local self = {value = value, left=nil, right=nil}
    self.add = Node.add
    self.contain = Node.contain
    self.length = Node.length
    self.height = Node.height
    return self
end

function Node:add(value)
    if value<self.value then
        if self.left == nil then
            self.left = Node.new(value)
            return true
        end
        return self.left:add(value)
    end

    if value>self.value then
        if self.right == nil then
            self.right = Node.new(value)
            return true
        end
        return self.right:add(value)
    end

    return false
end

function Node:contain(value)
    if value < self.value then
        if self.left == nil then
            return false
        end
        return self.left:contain(value)
    end

    if value > self.value then
        if self.right == nil then
            return false
        end
        return self.right:contain(value)
    end

    return true
end

function Node:length()
    local left_len = 0
    if self.left ~= nil then
        left_len = self.left:length()
    end

    local right_len = 0
    if self.right ~= nil then
        right_len = self.right:length()
    end

    return left_len + right_len + 1
end

function Node:height()
    local left_height = 0
    if self.left ~= nil then
        left_height = self.left:height()
    end

    local right_height = 0
    if self.right ~= nil then
        right_height = self.right:height()
    end

    return 1 + math.max(left_height, right_height)
end

File: BST/ruby/node.rb
class Node 
    @value
    @left
    @right

    def initialize(value)
        @value = value
    end

    def add(value)
        if value < @value
            if @left.nil?
               @left = Node.new(value)
               return true
            end
            return @left.add(value) 
        end
        if value > @value
            if @right.nil?
                @right = Node.new(value)
                return true
            end
            return @right.add(value)
        end
    end

    def contain(value)
        if value < @value
            if @left.nil?
               return false
            end
            return @left.contain(value) 
        end
        if value > @value
            if @right.nil?
                return false
            end
            return @right.contain(value)
        end
        return true
    end

    def length()
        a = @left.nil? ? 0 : @left.length()
        b = @right.nil? ? 0 : @right.length()
        return a + b + 1

    end

    def height()
        
        a = @left.nil? ? 1 : @left.height() + 1
        b = @right.nil? ? 1 : @right.height() + 1
        return a>b ? a : b
    end

end
